Blog post about the 5 years of Preact X (#1139)
preactjs/preact-www: content/en/blog/preact-x.md

---
title: Preact X, a story of stability
date: 2024-05-24
authors:
  - Jovi De Croock
---

# Preact X, a story of stability

A lot of you have been waiting for [Preact 11](https://github.com/preactjs/preact/issues/2621), announced in an issue opened
way back in July 2020, and to be clear I was one of the most excited people for v11.
When we started thinking about Preact 11 we believed that there was no way to introduce the changes we had in mind
in Preact X without breaking changes, some of the things we had in mind:

- Using a backing VNode structure to reduce GC, by doing this we'd only use the result of `h()` to update our backing-node.
- Reconciler performance, adding fast paths for mounting/...
- Some changes like removing `px` suffixing, `forwardRef` and breaking IE11 support.

> - Addressing event/child diffing bugs.

Reviewer: Should we add the `keeping ref in props` change here. We had to backtrack to react compat.

The above were our goals, when breaking these down we can clearly see that there's a distinction here, only point 3 would
be a breaking change for our consumers while point 1 would have a huge impact to folks building extensions like Preact Signals,
Preact devtools, ... This could break a lot of the cool stuff our community is working on to extend Preact.

As we were [experimenting](https://github.com/preactjs/preact/tree/v11) we went a new type of diffing, named
[skew based diffing](https://github.com/preactjs/preact/pull/3388), we saw real performance
improvements as well as it fixing a bunch of long-running bugs. As time went on and we invested more time in
these experiments for Preact 11, we started noticing that the improvements we were landing didn't need to be exclusive to Preact 11.

## Releases

Since the aforementioned Preact 11 issue there have been 18 (!!) minor versions of Preact X.
Many of them have been directly inspired by work done on Preact 11. Let's go over a few and look at the impact.

### 10.5.0

> The introduction of [resumed hydration](https://github.com/preactjs/preact/pull/2754) this functionality basically allows suspending during

Reviewer 2: ```suggestion
The introduction of [resumed hydration](https://github.com/preactjs/preact/pull/2754) -- this functionality basically allows suspending during
```

Semicolon would work too I think, but needs something to break up the flow of the sentence.

### 10.19.0

In 10.19.0 Marvin applied his research from [fresh](https://fresh.deno.dev/) to add [pre-compiled JSX functions](https://github.com/preactjs/preact/pull/4177),
this basically allows you to pre-compile your components during transpilation, when render-to-string is running we just have to concatenate the strings rather
than allocating memory for the whole VNode tree. The transform for this is exclusive to Deno at the moment but the general concept is present in Preact!

### 10.20.2

We have faced a number of issues where an event could bubble up to a newly inserted VNode which would result in undesired behaviour, this was fixed
[by adding an event-clock](https://github.com/preactjs/preact/pull/4322). In the following scenario, you would click the button which sets state, the browser
interleaves event bubbling with micro-ticks, which is also what Preact uses to schedule updates. This combination means that Preact will update the UI, meaning
that the `<div>` will get that `onClick` handler which we'll bubble up to and invoke the `click` again toggling this state immediately off again.

```jsx
const App = () => {
  const [toggled, setToggled] = useState(false);

>   return toggled ? <div onClick={() => setToggled(false)}><span>clear</span></div> : <div><button onClick={() => setToggled(true)}>toggle on</button></div>

Reviewer: Nit: let's format this to be more readable

- [Streaming rendering](https://github.com/preactjs/preact-render-to-string/pull/354)
- [Prefresh](https://github.com/preactjs/prefresh)
- [The vite preset with pre-rendering](https://github.com/preactjs/preset-vite#prerendering-configuration)
- [A new async router](https://github.com/preactjs/preact-iso)
- [Create Preact](https://github.com/preactjs/create-preact)

We value our ecosystem and we value the extensions being built through our [`options API`](https://marvinh.dev/blog/preact-options/),
this is one of the main drivers behind not wanting to introduce these breaking changes but instead, allow you all to benefit
from our research without a painful migration path.

This doesn't mean that Preact 11 won't happen but it might not be the thing that we initially thought it would be, instead we might just drop IE11 support and give you
those performance improvments, all while giving you the stability of Preact X. We could instead work on a meta-framework for Node to accompany Deno as a great
way to write Preact, we could bet more on signals... There are a lot of options while avoiding leaving a part of the community behind.

> We aren't gone, we are here to listen to you and we want to grow this community together.

Reviewer: I don't have a better suggestion for the phrasing yet, but this reads a bit odd to me.

1. Worried that folks think that Preact is mostly for Deno going onwards

> There are a lot of options while avoiding leaving a part of the community behind.
This feels like our main worry is not to be left behind? Do we feel that way?

---
title: Preact X, a story of stability
date: 2024-05-24
authors:
  - Jovi De Croock
---

# Preact X, a story of stability

A lot of you have been waiting for [Preact 11](https://github.com/preactjs/preact/issues/2621), announced in an issue opened
way back in July 2020, and to be clear I was one of the most excited people for v11.
When we started thinking about Preact 11 we believed that there was no way to introduce the changes we had in mind
in Preact X without breaking changes, some of the things we had in mind:

- Using a backing VNode structure to reduce GC, by doing this we'd only use the result of `h()` to update our backing-node.

> - Reconciler performance, adding fast paths for mounting/...

Reviewer: Minor rewording:

```suggestion
- Reconciler performance, to allow optimized paths for mounting/unmounting/etc
```

- Some changes like removing `px` suffixing, `forwardRef` and breaking IE11 support.
- Keeping ref in props.
- Addressing event/child diffing bugs.

> The above were our goals, when breaking these down we can clearly see that there's a distinction here, only point 3 would
> be a breaking change for our consumers while point 1 would have a huge impact to folks building extensions like Preact Signals,
> Preact devtools, ... This could break a lot of the cool stuff our community is working on to extend Preact.

Reviewer: ```suggestion
These were our initial goals for v11, but upon going down this path, we realised that many of those changes weren't actually breaking changes and could be released directly in v10 in a non-breaking way. Only the third point, removing the `px` suffix and passing `ref` directly in props as well as dropping IE11, fall into the breaking changes category. We went with releasing the other features in the stable v10 release line, which allows any Preact user to benefit from them immediately, without having to change their code.

Preact has a much bigger userbase today compared to when we made the original plans for v11. It enjoys wide usage in many small to big companies for mission critical software. We really want to be sure that any breaking changes we may introduce, are absolutely worth the cost of moving the whole ecosystem over to it.
```

Reviewer 2: That last sentence in this suggestion should drop the comma I think, breaks the flow.

Reviewer: Good point, updated.

downloads in the past week, the last 8.x release was 5 years ago, just to show that a part of the community gets left behind.

Stability is great, we as the Preact team love stability. We actually released multiple major features on other ecosystem projects:

- [Signals](https://github.com/preactjs/signals)
- [Async rendering](https://github.com/preactjs/preact-render-to-string/pull/333)
- [Streaming rendering](https://github.com/preactjs/preact-render-to-string/pull/354)
- [Prefresh](https://github.com/preactjs/prefresh)
- [The vite preset with pre-rendering](https://github.com/preactjs/preset-vite#prerendering-configuration)
- [A new async router](https://github.com/preactjs/preact-iso)
- [Create Preact](https://github.com/preactjs/create-preact)

We value our ecosystem and we value the extensions being built through our [`options API`](https://marvinh.dev/blog/preact-options/),
this is one of the main drivers behind not wanting to introduce these breaking changes but instead, allow you all to benefit
from our research without a painful migration path.

> This doesn't mean that Preact 11 won't happen but it might not be the thing that we initially thought it would be, instead we might just drop IE11 support and give you
> those performance improvments, all while giving you the stability of Preact X. We could instead work on a meta-framework for Node like [Fresh](https://fresh.deno.dev/) as a great
> way to write Preact, we could bet more on signals... There are a lot of options while avoiding leaving a part of the community behind.

Reviewer: Minor rewording.

```suggestion
This doesn't mean that Preact 11 won't happen but it might not be the thing that we initially thought it would be. Instead, we might just drop IE11 support and give you
those performance improvements, all while giving you the stability of Preact X. There are many more ideas floating around and we're very interested in the wider Preact experience in the context of meta-frameworks that provide things like routing out of the box. We're exploring this angle in our vite preset as well as [Fresh](https://fresh.deno.dev/) to get a good feel what a Preact first meta framework should look like.
```
